content/source/docs/enterprise/sentinel/import/index.html.md

```diff
@@ -24,100 +24,178 @@ see [Mocking Terraform Sentinel Data](../mock.html).
 
 -> **Note:** Terraform Enterprise does not currently support custom imports.
 
-## Useful Idioms for Terraform Sentinel Policies
+## Useful Functions and Idioms for Terraform Sentinel Policies
 
-Terraform's internal data formats are complex, which means basic Sentinel policies for Terraform are more verbose than basic policies that use simpler data sources.
-
-This will improve in future versions of Terraform and Sentinel; in the meantime, be aware of the following idioms as you start writing policies for Terraform.
+The following functions and idioms will be useful as you start writing Sentinel policies for Terraform.
 
 ### To Find Resources, Iterate over Modules
 
-The most basic Sentinel task for Terraform is to enforce a rule on all resources of a given type. Before you can do that, you need to get a collection of all the relevant resources.
-
-The easiest way to do that is to copy a function like the following into any policies that examine every resource in a configuration:
+The most basic Sentinel task for Terraform is to enforce a rule on all resources of a given type. Before you can do that, you need to get a collection of all the relevant resources. The easiest way to do that is to copy a function like the following into your policies:
 
 ```python
 import "tfplan"
-
-# Get an array of all resources of the given type (or an empty array).
-get_resources = func(type) {
-	if length(tfplan.module_paths else []) > 0 { # always true in the real tfplan import
-		return get_resources_all_modules(type)
-	} else { # fallback for tests
-		return get_resources_root_only(type)
-	}
-}
-
-get_resources_root_only = func(type) {
-	resources = []
-	named_and_counted_resources = tfplan.resources[type] else {}
-	# Get resource bodies out of nested resource maps, from:
-	# {"name": {"0": {"applied": {...}, "diff": {...} }, "1": {...}}, "name": {...}}
-	# to:
-	# [{"applied": {...}, "diff": {...}}, {"applied": {...}, "diff": {...}}, ...]
-	for named_and_counted_resources as _, instances {
-		for instances as _, body {
-			append(resources, body)
-		}
-	}
-	return resources
-}
-
-get_resources_all_modules = func(type) {
-	resources = []
-	for tfplan.module_paths as path {
-		named_and_counted_resources = tfplan.module(path).resources[type] else {}
-		# Get resource bodies out of nested resource maps, from:
-		# {"name": {"0": {"applied": {...}, "diff": {...} }, "1": {...}}, "name": {...}}
-		# to:
-		# [{"applied": {...}, "diff": {...}}, {"applied": {...}, "diff": {...}}, ...]
-		for named_and_counted_resources as _, instances {
-			for instances as _, body {
-				append(resources, body)
-			}
-		}
-	}
-	return resources
+import "strings"
+
+# Find all resources of a specific type from all modules using the tfplan import
+find_resources_from_plan = func(type) {
```

Comment:
I still would really prefer the logic for `get_instance_address` be rolled into this function so that the pattern that we promote returns the full resource address instead of having to dip into it further.

I get that this differs between imports, so maybe if we provide some sort of "tabbed" code sample block that would be better. @nfagerlund do we have precedent of this happening in other places currently and can we do that here?

Reply:
Getting the full address of resource instances in find_resource_from_plan would require iterating over all the resources which is not necessary at this point.  What I am doing is returning a map of maps, one per module, for which each key is the joined_path of the module.  Iterating over all the resources and their instances inside the find_resources_from_plan function would make it much more complex than it needs to be.

You had previously objected to my using path as the keys of the resource_maps even though the code worked fine. I know that https://docs.hashicorp.com/sentinel/language/spec#map-literals says that keys of maps are supposed to be boolean, numeric, or string, but my code did work fine.  I actually liked the way the code looked better when I did set the key to path and computed the full resource instance address in the get_instance_address function.  While that required rederiving the joined path more often and is therefore less efficient, it did keep all the code for the derivation of the address in a single function.  I would be happy to revert back to that pattern.  I think efficiency of the Sentinel code is less important than clear organization.

If you eventually end up writing standard functions in go that do what my functions are doing, then I would write my policies to use those functions.  But since we currently have to paste the contents of these functions inside the policies, I don't want to make them more complicated than necessary.

Reply:
> Iterating over all the resources and their instances inside the find_resources_from_plan function would make it much more complex than it needs to be.

Aside from the extra two loops and a bit of extra details on the `join`, what is more complex about the below example?

```
find_resources_by_type = func(type) {
	result = {}

	for tfplan.module_paths as path {
		for tfplan.module(path).resources[type] else {} as name, by_index {
			for by_index as index, r {
				result[strings.join([module_address(path), type, name, index], ".")] = r
			}
		}
	}

	return result
}
```

(Where `module_address` is a helper function that returns a properly joined module address.)

> I actually liked the way the code looked better when I did set the key to path and computed the full resource instance address in the get_instance_address function. While that required rederiving the joined path more often and is therefore less efficient, it did keep all the code for the derivation of the address in a single function.

I still don't see how this can be useful aside from niche uses.

Returning a map of resources keyed on the full resource path allows a very simple path to reporting a resource in violation - printing the key. It also allows for much easier iteration as you have already done the work of looping on the two inner maps.

As far as I can see, the rest of the example doesn't demonstrate any value in providing less granularity than the final set of actual resources keyed on address - nor do I think it should. In fact, `validate_instance_types` could be greatly simplified if it was operating on a flat set of resources versus having to deeply iterate and combine the resource address itself.